### PR TITLE
Code CLI - Add --state filter to --list-extensions

### LINKED ISSUE #19083: Code command line add a --list-enabled-extensions
Currently you can only list all extensions. It would be useful to only list the enabled extensions.

### PR DESCRIPTION
**Pull request by lszomoru**
This PR fixes #19083 

With this change I have added the `--state` filter to the `--list-extensions` command. The filter only uses global enablement state (extension is globally enabled/disabled). In order to use the global enablement service I had to connect to the SQLLite database to get the list of disabled extensions. Not sure if that is the right approach here so any guidance would be highly appreciated. Thanks! 

Usage:
`code --list-extensions --state enabled` - will list all extensions that are enabled
`code --list-extensions --state disabled` - will list all extensions that are disabled

### CODE REVIEW

**Comment by lszomoru** [comment]
tagging @joaomoreno, @bpasero, @sandy081 to sign off on the change

**Review by bpasero — COMMENTED** [review]
I talked with @joaomoreno about this earlier today. We are using SQLite as a DB for storing state. Normally things should be OK accessing the DB from CLI if no other process is running. However, in case VSCode is already running, we now might have 2 processes connecting to the same DB on disk. We do not run SQLite in a mode that locks the DB for single-process use, but I fear this may have a chance to cause corruption if not carefully understood how SQLite behaves in this case. 

Couple of thoughts:
* can we provide an implementation of SQLite state access that is guaranteed to not brick an instance that is accessed meanwhile from another process (e.g. with some options)
* could the CLI process first check if it is the first one or not and then delegate the query to the first instance?

I would probably go with the first option and implement a readonly version of [`SQLiteStorageDatabase`](https://github.com/microsoft/vscode/blob/cd0d6485221e0f0e89d04aa760b6110db925575f/src/vs/base/parts/storage/node/storage.ts#L31). Still, we need to understand how SQLite behaves.

Sorry, https://github.com/microsoft/vscode/issues/19083 should have not been marked as help wanted so easily.

**Comment by sandy081** [comment]
I would go for second option 

> could the CLI process first check if it is the first one or not and then delegate the query to the first instance?

Because there are other use case to update storage from CLI - https://github.com/microsoft/vscode/issues/63903

Check the same suggestion from @joaomoreno here - https://github.com/microsoft/vscode/issues/63903#issuecomment-470836331

**Comment by lszomoru** [comment]
no worries about the issue being marked as "help wanted", and thanks for the feedback.

Implementing a "read-only" option for `SQLiteStorageDatabase` seems a reasonable option which would unblock the completion of this issue however as @sandy081 calls it out there are several other scenarios for which the CLI needs write access. If we were to add storage support for the CLI we should do it in a way that fits both read and write scenario so @joaomoreno's suggestion makes sense to me especially that there is already an established pattern. I can take a stab at it and see how far I can get.

Is there any particular reason why we open a connection to the SQLLite database and keep it open for the lifetime of the process? Is there such a big overhead for establishing the connection? I am asking as another option which would save us from the interprocess communication would be for each process to open the connection/execute query/close the connection. I am also reviewing the [SQLLite documentation](https://sqlite.org/lockingv3.html) for concurrency/locking.

**Comment by joaomoreno** [comment]
> Is there any particular reason why we open a connection to the SQLLite database and keep it open for the lifetime of the process? Is there such a big overhead for establishing the connection? I am asking as another option which would save us from the interprocess communication would be for each process to open the connection/execute query/close the connection. I am also reviewing the [SQLLite documentation](https://sqlite.org/lockingv3.html) for concurrency/locking.

This would definitely be way too far, just to tackle this issue.

I would even argue read/write access is taking it too far. Yes, it would be great to have, but no feature is too pressing to force this amount of change.

I would investigate on having concurrent read-only access to the db, which after a few web searches, definitely seems like a feature SQLite supports. This would not only fix this feature, unblock others and would keep things simple.

**Comment by sandy081** [comment]
Agreed that implementing read & write is too much for this issue. But there are other issues which need write access too. 

- Ability to enable / disable from CLI - https://github.com/microsoft/vscode/issues/52639#issuecomment-422797554
- #63903

If we started thinking and investing on this, is not it recommended to think about the complete solution? I do not understand if the complete solution is technically feasible or not.

Also this will make storage service inconsistent  (there by other services like extension enablements too) across the processes we have.

**Comment by lszomoru** [comment]
As the issue in question is 3 years old, I do not believe that there is an incentive for a tactical change in order to implement the requested feature. In light of that, I would like to advocate for the right solution which will unblock all request that are currently blocked due to the fact that the CLI does not have storage access. 

Given that there is already a pattern for delegating storage queries, I will go ahead and explore that. At a later date we can explore keeping the connection open just for the time of executing the query since that seems to be the best solution long term.

**Comment by lszomoru** [comment]
I had a quick chat with @joaomoreno earlier today and we have agreed that for the scope of this particular issue we will go with adding the capability to create a read-only connection to the SQLite database. This works as expected however there is one edge case where this will not work. If the global SQLite database does not exist and the CLI is run that it will fail as we cannot run the script to create the database when we are using a read-only connection. If you feel that we should address this edge case, we should discuss further. Thanks!

**Comment by bpasero** [comment]
@lszomoru are you able to resolve the conflicts? sorry, I pushed a lot of refactorings around environment

**Comment by lszomoru** [comment]
@bpasero, sure. Let me resolve the conflicts and update the pull request.

**Comment by lszomoru** [comment]
@bpasero I have resolved the conflicts and updated the pull request.

File: src/vs/base/parts/storage/node/storage.ts
Comment (bpasero) [inline src/vs/base/parts/storage/node/storage.ts:23]: I would prefer if we simply had the option to open the DB as `readonly` and not expose the underlying SQLite modes as option. I think in the end we just have 1 client that wants to open the DB readonly so I would optimise for that scenario.

File: src/vs/base/parts/storage/node/storage.ts
Comment (bpasero) [inline src/vs/base/parts/storage/node/storage.ts:38]: I would prefer if we do not use any default mode but leave that up to the SQLite implementation. We only set the mode to readonly when instructed, but otherwise leave it as is.

File: src/vs/base/parts/storage/node/storage.ts
Comment (bpasero) [inline src/vs/base/parts/storage/node/storage.ts:302]: As mentioned above, pass `undefined` instead of default mode to leave the default up to the underlying library.

File: src/vs/base/parts/storage/node/storage.ts
Comment (bpasero) [inline src/vs/base/parts/storage/node/storage.ts:309]: You are missing a return here and as such resolve the promise twice.

File: src/vs/code/node/cliProcessMain.ts
Comment (bpasero) [inline src/vs/code/node/cliProcessMain.ts:142]: `await` not needed here

File: src/vs/code/node/cliProcessMain.ts
Comment (bpasero) [inline src/vs/code/node/cliProcessMain.ts:328]: Extract to a place where it can be reused?

File: src/vs/code/node/cliProcessMain.ts
Comment (bpasero) [inline src/vs/code/node/cliProcessMain.ts:347]: This is not needed as `flush` is only used for writing

File: src/vs/code/node/cliProcessMain.ts
Comment (bpasero) [inline src/vs/code/node/cliProcessMain.ts:342]: This deserves a bit more documentation why the DB can only be opened readonly.

**Review by bpasero — CHANGES_REQUESTED** [review]
A bit of feedback, I focused on the DB pieces 👍

**Comment by sandy081** [comment]
Discussed with @joaomoreno and @lszomoru and I do not think this is worth adding this feature for the amount of work that is done and the debt it might create.

Also the solution might not be complete.. we also have workspace enabled extensions. So showing the global enablement might not be complete enough
and adding this is asking for more trouble.

**Pull request closed without merging**
